src/NServiceBus.RabbitMQ/Receiving/MessageConverter.cs

```diff
@@ -98,50 +97,65 @@ string DefaultMessageIdStrategy(BasicDeliverEventArgs message)
 
         static string ValueToString(object value)
         {
-            var s = value as string;
-            if (s != null)
-            {
-                return s;
-            }
-
             var bytes = value as byte[];
             if (bytes != null)
             {
                 return Encoding.UTF8.GetString(bytes);
             }
 
-            var dictionary = value as IDictionary<string, object>;
+            var dictionary = value as Dictionary<string, object>;
             if (dictionary != null)
             {
-                var valuesToJoin = new List<string>();
+                var sb = new StringBuilder();
 
-                // ReSharper disable once LoopCanBeConvertedToQuery
                 foreach (var kvp in dictionary)
                 {
-                    valuesToJoin.Add(string.Concat(kvp.Key, "=", ValueToString(kvp.Value)));
+                    sb.Append(kvp.Key);
+                    sb.Append("=");
+                    sb.Append(ValueToString(kvp.Value));
+                    sb.Append(",");
                 }
 
-                return string.Join(",", valuesToJoin);
+                if (sb.Length > 0)
+                {
+                    sb.Remove(sb.Length - 1, 1);
+                }
+
+                return sb.ToString();
             }
```

Comment:
Cool, yeah that does seem like the minimum change that would make everything work, though from my actual workflow standpoint, I added the timestamp stuff last.

BTW, one of the additional reasons the commits ended up the way they did, and not more individual work commits, is that I originally started this work on the native delays branch and had some stuff specific to in mixed in. I also added tests at different stages as well. There was a decent amount of interactive rebasing involved to get everything over on develop cleanly!